File: cpp/day04_scratchcards.cpp
// Day 4. Scratchcards
// https://adventofcode.com/2023/day/4
#include<iostream>
#include<fstream>
#include<set>
#include<cmath>

using namespace std;

struct ScratchCard {
    int n;
    set<int> winningnums;
    set<int> mynums;
};

void print_card(ScratchCard* card) {
    cout << "Card: " << card->n << endl << "Winning Nums: ";
    for (auto n : card->winningnums) {
        cout << n << ' ';
    }
    cout << endl;
    cout << "My Nums: ";
    for (auto n : card->mynums) {
        cout << n << ' ';
    }
    cout << endl;
}


set<int> parse_nums(string s, string delimiter = " ") {
    set<int> res;
    size_t pos = 0;
    string next;
    while ((pos = s.find(delimiter)) != string::npos) {
        next = s.substr(0, pos);
        if (next.length() != 0) {
            res.insert(stoi(next));
        }
        s.erase(0, pos + delimiter.length());
    }
    // last item
    if (s.length() > 0) res.insert(stoi(s));
    return res;
}





auto read_input() {
    string filename = "../inputs/day04_input.txt";
    cout << "Input file: " << filename << endl;
    string line;
    vector<ScratchCard*> cards;
    ifstream input(filename);
    while (getline(input, line)) {
        ScratchCard* c = new ScratchCard();
        int idx0 = line.find(" ");
        int idx1 = line.find(':');
        int idx2 = line.find('|');
        c->n = stoi(line.substr(idx0+1, idx1-idx0));
        c->winningnums = parse_nums(line.substr(idx1+2, idx2-idx1-2));
        c->mynums = parse_nums(line.substr(idx2+2));
        cards.push_back(c);
    }
    input.close();
    return cards;
}


int get_wins(ScratchCard* card) {
    set<int> wins;
    set_intersection(card->mynums.begin(), card->mynums.end(),
                    card->winningnums.begin(), card->winningnums.end(),
                    inserter(wins, wins.begin()));
    return wins.size();
}

int score_card(ScratchCard* card) {
    int wins = get_wins(card);
    if (wins == 0) return 0;
    return pow(2, wins-1);
}


auto part1() {
    vector<ScratchCard*> cards = read_input();

    int score = 0;
    for (ScratchCard* card : cards) {
        int s = score_card(card);
        // print_card(card);
        // cout << s << endl;
        score += s;
    }
    return score;
}


auto part2() {
    vector<ScratchCard*> cards = read_input();
    vector<int> counts;
    for (int i=0; i<cards.size(); i++) {
        counts.push_back(1);
    }

    ScratchCard* card;
    int wins;
    int currcount;
    int totalcount = 0;
    for (int i=0; i<cards.size(); i++) {

        card = cards[i];
        currcount = counts[i];
        wins = get_wins(card);

        // Add current count
        totalcount += currcount;

        // increase cards
        for (int j=0; j<wins; j++) {
            counts[i+j+1] += currcount;
        }
    }

    return totalcount;
}


int main() {
    auto p1 = part1();
    cout << "Part 1: " << p1 << endl;

    cout << "--------------" << endl;

    auto p2 = part2();
    cout << "Part 2: " << p2 << endl;
}

File: python/day04_scratchcards.py
# Day 4: Scratchcards
# https://adventofcode.com/2023/day/4

from util.input_util import read_input_file
from collections import deque

class ScratchCard:
    def __init__(self, card:int, winNums:set, myNums:set):
        self.card = card
        self.winNums = winNums
        self.myNums = myNums
        self.wins = self.myNums.intersection(self.winNums)
    
    def score(self) -> int:
        n = len(self.wins)
        if n == 0: return 0
        return 2 ** (n-1)

    def getWins(self) -> set:
        return self.wins
    
    def getWinsCount(self) -> int:
        return len(self.wins)
    
    def getNum(self) -> int:
        return self.card
        
    def __repr__(self):
        return f"CARD: {self.card}\nWINNING: {self.winNums}\nMY: {self.myNums}\nWINS: {self.wins}"
        
        

def parse_lines():
    lines = read_input_file(4)
    cards = {}
    for l in lines:
        c, l = l[:l.find(':')], l[l.find(':')+2:]
        c = int(c.split()[1])
        win, my = l.split(' | ')
        win = set([int(x) for x in win.split()])
        my = set([int(x) for x in my.split()])
        card = ScratchCard(c, win, my)
        cards[c] = card
        
    return cards




def solution1():
    cards = parse_lines()
    totalscore = 0
    for _, card in cards.items():
        totalscore += card.score()
    
    return totalscore
        
    
    
def solution2():
    cards = parse_lines()
    counts = {c:1 for c in cards}
    totalcards = 0
    
    for c, card in cards.items():
        n = counts[c]
        totalcards += n

        for i in range(card.getWinsCount()):
            counts[c+i+1] += n
        
    return totalcards
        
    
    
    
    
if __name__ == '__main__':
    print("Part 1:", solution1())
    
    print('--------------')
    
    print("Part 2: ", solution2())